src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -175,8 +191,9 @@ internal void Freeze()
             // Sort top-level types.
             _orderedTopLevelTypes = _synthesizedTopLevelTypes.OrderBy(kvp => kvp.Key).Select(kvp => (Cci.INamespaceTypeDefinition)kvp.Value).AsImmutable();
 
-            // Sort proxy types.
-            _orderedProxyTypes = _proxyTypes.OrderBy(kvp => kvp.Key.Size).ThenBy(kvp => kvp.Key.Alignment).Select(kvp => kvp.Value).AsImmutable();
+            // Sort nested types.
+            _orderedNestedTypes = _dataFieldTypes.OrderBy(kvp => kvp.Key.Size).ThenBy(kvp => kvp.Key.Alignment).Select(kvp => kvp.Value).OfType<ExplicitSizeStruct>()
```

Comment:
>.OfType<ExplicitSizeStruct>() [](http://example.com/codeflow?start=136&length=29)

The purpose of ```OfType``` API is to filter the input by type. I do not think that we intend to do that here. If our goal is to do a simple type cast, there is a ```Cast``` helper, but, I think it would be more efficient to perform the cast in the lambda passed to the preceding ```Select``` call.

Reply:
Would it make sense to strengthen the type of values stored in ```_dataFieldTypes``` to ```ExplicitSizeStruct``` or to ```Cci.INestedTypeDefinition```?

Reply:
`_dataFieldTypes` might contain System.Int* types, not just ExplicitSizeStructs. The filtering via OfType was always performed - it has just moved from here: [https://github.com/dotnet/roslyn/blob/98ea496177bc8e607dbf454abd6b5a5e4678aed2/src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs#L414](https://github.com/dotnet/roslyn/blob/98ea496177bc8e607dbf454abd6b5a5e4678aed2/src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs#L414)